editor/components/block-list/block.js

```diff
@@ -348,6 +354,96 @@ export class BlockListBlock extends Component {
 	}
 
+	onDragStart( event ) {
+		const dragInset = document.getElementById( `block-drag-inset-${ this.props.uid }` );
+		const block = document.getElementById( `block-${ this.props.uid }` );
+
+		// Closure to remove the cloned node from the DOM (fired within timeout below)
+		const removeBlockClone = ( _blockList, _cloneWrapper ) => {
+			return () => {
+				_blockList.removeChild( _cloneWrapper );
+			};
+		};
+
+		// Closure to hide the visible block and show inset in its place (fired within timeout below)
+		const showInset = () => {
+			return () => {
+				this.setState( state => Object.assign( {}, state, { dragging: true } ) );
+				document.body.classList.add( 'dragging' );
+			};
+		};
+
+		event.dataTransfer.setData(
+			'text',
+			JSON.stringify( {
+				uid: this.props.uid,
+				fromIndex: this.props.order,
+				type: BLOCK_REORDER,
+			} )
+		);
+
+		if ( 'function' === typeof event.dataTransfer.setDragImage ) {
+			const blockList = block.parentNode;
+			const cloneWrapper = document.createElement( 'div' );
+			const clone = block.cloneNode( true );
+			const blockRect = block.getBoundingClientRect();
+
+			// 1. Clone the current block, style it, and spawn over original block.
+
+			clone.id = `clone-${ block.id }`;
+
+			cloneWrapper.classList.add( 'editor-block-list__block-clone' );
+			// 40px padding for the shadow.
+			cloneWrapper.style.width = `${ blockRect.width + 40 }px`;
+			// Position clone right over the original block.
+			// 20px padding for the shadow.
+			cloneWrapper.style.top = `${ parseInt( blockRect.top, 10 ) - 20 }px`;
+			cloneWrapper.style.left = `${ parseInt( blockRect.left, 10 ) - 20 }px`;
+
+			cloneWrapper.appendChild( clone );
+			blockList.appendChild( cloneWrapper );
+
+			// 2. Set clone as drag image and remove from DOM:
+
+			// Display the drag image right over the block:
+			//   - Coordinates relative to the block and the cursor/
+			//   - Optimisation: if top ends of the block are outside the viewport,
+			//     then revert to top 0 (relative to cursor).
+			const top = parseInt( blockRect.top, 10 ) > 0 ?
+				parseInt( event.clientY, 10 ) - parseInt( blockRect.top, 10 ) + 20 :
+				parseInt( event.clientY, 10 ) + 20;
+
+			event.dataTransfer.setDragImage(
+				cloneWrapper,
+				parseInt( event.clientX, 10 ) - parseInt( blockRect.left, 10 ) + 20,
+				top
+			);
+
+			setTimeout( removeBlockClone( blockList, cloneWrapper ) );
+		}
+
+		// Hide the visible block and show inset in its place.
+		setTimeout( showInset() );
+	}
+
+	onDragEnd() {
+		const block = document.getElementById( `block-${ this.props.uid }` );
+		const dragInset = document.getElementById( `block-drag-inset-${ this.props.uid }` );
+
+		const resetBlockDisplay = () => {
+			return () => {
+				this.setState( state => Object.assign( {}, state, { dragging: false } ) );
+				document.body.classList.remove( 'dragging' );
+			};
+		}
+
+		setTimeout( resetBlockDisplay(), 0 );
+	}
+
+	onDrop( uid, toIndex ) {
```

Comment:
Why is `uid` passed as an argument here and in other methods like `onDragEnd` we use the one from `this.props.uid`?

Reply:
This is quite a complicated call and entirely different from the others actually. It goes through the dropzone provider, then the block dropzone component, and then here. So it's not really an event handler per se, while the others are.

Reply:
👍